### DISCUSSION TITLE
Documentation Updates - Querying a user's collection

### DISCUSSION THREAD

**Discussion by prcutler** [joalla/discogs_client#67] [General] [answered]
Hi,

Just starting a new thread after posting in #64 

I'm still fairly new to Python and reading developer documentation is sometimes still confusing.  As a quick introduction, I use to write a lot of use documentation for the GNOME desktop on Linux, but not a lot of developer docs. Writing docs is how I learned Linux and I find writing it down helps me learn, and more importantly, retain it.  But enough about me!

@AnssiAhola asked what kind of app I'm building.  I'm big into music and have a large record collection (as most people probably using this API!) In the first phase, I'm building a FastAPI web app that will query my collection via the Discogs API and show me a random record to play when I can't decide.  I'm also adding the ability to choose a specific record and the web app will in both cases display the album art, some info about the release, etc.

In the second phase, I'm going to connect the app to MQTT and it will send a message to an Adafruit LED Matrix that will display the album art.  My record player and stereo are in a different room on the opposite side of the wall from my home office, so I don't have to walk too far to flip a record, but I thought it would be cool to see the album art digitally.

This is all just a way for me to have a hobby and keep learning Python.   :)

With all that said, I'll be asking some pretty basic questions and I'll fork the repo and start working on some docs while I'm learning.  

Thanks in advance for anyone taking the time to help answer them!

--Paul

**Comment by JOJ0** [comment]
Hi, thanks from my end as well. As you noticed already some basics are missing in our docs. Simple examples about iterating through the collection woul indeed be a good start. Somethin like this would do: https://github.com/joalla/discogs_client/discussions/64#discussioncomment-1555587

**Reply by prcutler** [reply]
Thanks - I'm started reading through that last night as well, that was helpful in instantiating the identity and passing my token.  Still have a ways to go.  :)

**Comment by prcutler** [comment]
So I'm replacing one of the methods I've written with one using the API instead.  I want to query my collection (folder 0), have it pick a random release, and return the release ID.

How do I parse a paginated list?  My function looks like this:

```    def get_folder_count2():
            lp_count = len(discogs_data.identity().collection_folders[0].releases)
            print(lp_count)
            
            random_lp = random.randint(0, lp_count)
            print("Random # = ", random_lp)
            
            random_album_release_id = discogs_data.identity().collection_folders[0].releases.id
            print("Random_ID = ", random_album_release_id)
            
            return random_album_release_id, 0
```

And returns this:

```
820
Random # =  615
Random_ID =  <discogs_client.models.PaginatedList object at 0x10b745780>
(<discogs_client.models.PaginatedList object at 0x10b745780>, 0)
```

So I have a `PaginatedList` object, but the only search result I get in the docs is about doing a [search that returns a list of objects](https://python3-discogs-client.readthedocs.io/en/latest/discogs_client.html?highlight=pagination#discogs_client.client.Client.search), but I'm not sure how to get data out of it.

Thinking it was a list I tried this instead, but got a 404:  `random_album_release_id = discogs_data.identity().collection_folders[0].releases.id[820]`

I'm guessing some of my questions are more about still learning Python, but any pointers are appreciated.

Thanks!

Paul

**Reply by JOJ0** [reply]
I'll take a closer look later. Cooking atm and on mobile ;-)

But one thing I noticed: You can only iterate through the ....collfolders[0].releases

Not pssible to pass an id index to .releases as you did. Have a look at the example from the other thread I mentioned above again. 

HTH so far, sorry for being brief!

**Comment by JOJ0** [comment, marked as answer]
Hi, sorry I was wrong, you _can_ pass an id but to the .releases already:

```
>>> my_client.identity().collection_folders[0].releases[124].release.title
'New Dawn / Cat & Mouse'
>>> 
>>> 
>>> my_client.identity().collection_folders[0].releases[124].release
<Release 11117 'New Dawn / Cat & Mouse'>
>>> 
>>> my_client.identity().collection_folders[0].releases[124].release.id
11117
>>> dir(my_client.identity().collection_folders[0].releases[124].release)
['__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_known_invalid_keys', 'artists', 'changes', 'client', 'companies', 'country', 'credits', 'data', 'data_quality', 'delete', 'fetch', 'formats', 'genres', 'id', 'images', 'labels', 'marketplace_stats', 'master', 'notes', 'refresh', 'save', 'status', 'styles', 'thumb', 'title', 'tracklist', 'url', 'videos', 'year']
>>> 
```

So this is kind of the id in the paginated list I guess. This could be your random number.

Use the dir() function to find out how to see what attributes are possible.

**Reply by prcutler** [reply]
Thanks for all the quick replies!  I see what I was doing wrong with `dir` now.  This is super helpful and one method just went from 40 lines of code to about 5.

**Reply by JOJ0** [reply]
Nice! Have fun moving on! Very nice project idea btw!

**Comment by prcutler** [comment]
Your help got me on the right track and I was able to refactor about half my code today using the API.  I have a few notes filled with API calls to start writing some docs around it that might help some folks in the future, too.

I think the only question I have at this point is why release.artists returns a list of one item, but includes the artist ID and name:

`print(my_data.release(20017387).artists, my_data.release(20017387).title)`

Returns:

`[<Artist 2953514 'Chvrches'>] Screen Violence` - I can parse it programatically, but was curious.  

Thanks again and I'll have a PR in the next few days after I update a couple more things.

Paul

**Reply by JOJ0** [reply]
I think artists is a list because potentially a record could be produced by more than one artists. What you see when you print `artists` directly, is the `__repr__` of the python object. All the discogs_client objects have suche a short "representation" (?) string coded into them. When you dig deeper by giving the id, like `artists[0]` you can get the actual details of the first artist listed on discogs for this release, like artist name, artist id, artist page links, etc. (use `dir()` again to find out). I guess it's better the api and thus the discogs client stays consitent here and always give artists as a list, no matter if it's one or multiple artists.

**Reply by JOJ0** [reply]
Here is an example of multiple artists and how discogs saves it:

The record is by an artist called "Beta 2 & Zero Tolerance", actually they are two guys and usually work together, but on the B-Side there is a track from "Zero Tolerance" alone but this time collaborating with a guy called "Naptha". Looks pretty confusing at first look:
```
>>> my_client.identity().collection_folders[0].releases[124].release.artists[1].name
'Zero Tolerance'
>>> my_client.identity().collection_folders[0].releases[124].release.artists[2].name
'Naphta'
>>> my_client.identity().collection_folders[0].releases[124].release.artists[0].name
'Beta 2 & Zero Tolerance'
>>> my_client.identity().collection_folders[0].releases[124].release.artists[3].name
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
IndexError: list index out of range
>>> 
```


But when you look at the release with all the context, it actually makes perfectly sense: https://www.discogs.com/release/11117-Beta-2-Zero-Tolerance-Zero-Tolerance-vs-Naphta-New-Dawn-Cat-Mouse

**Comment by prcutler** [comment]
Thanks again - that makes a lot more sense.  I was looking at a couple records in my collection the wrong way.  For example, I thought The Traveling Wilburys might list the individual artists, but it doesn't.  But now that you point out that example, I see it and with records in my collection like https://www.discogs.com/master/1316036-Poli%C3%A7a-And-Stargaze-Music-For-The-Long-Emergency.  I also didn't realize I could call `.name` on the API call.  Need to remember to always test with `dir()`!

Once again, thanks for the quick reply!

**Comment by prcutler** [comment]
One last question (I hope):

For the master release, is it possible to query the year?

If I do a dir on `print(dir(my_data.master))`:

```
['__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', 'artists', 'client', 'credits', 'data', 'duration', 'fetch', 'position', 'title']
```
Looking at that last line of artists through title, I'm guessing not?

I can do a query on the Master ID from the Client:

print(my_data.release(20017387).master.id) and get back the Master ID of 2272042, but `.year` doesn't work. 

I like to get the master release year because reissues year field are the date of the reissue.

What am I missing this time?  :) 

Thanks in advance (again).

--P

**Reply by AnssiAhola** [reply]
The `Master` model does not seem to have a `year` property https://github.com/joalla/discogs_client/blob/80eae2522606603ee88a7ad14900d5f2402ebc72/discogs_client/models.py#L535-L553
That needs to be added for sure 😅 
Meantime you can use `release(9894023).master.fetch('year')`

**Comment by prcutler** [comment]
Thanks for the clarification!  I need to dig into `fetch` - I hadn't even noticed that one. 

Once again, thank you.
